File: leetcode/_133.java
import java.util.*;

class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

public class _133 {
    public Node cloneGraph(Node node) {
        Map<Node, Node> map = new HashMap<>();
        if (node == null) {
            return null;
        }
        Node newNode = new Node(node.val);
        map.put(node, newNode);

        Queue<Node> queue = new ArrayDeque<>();
        queue.offer(node);
        while (!queue.isEmpty()) {
            Node currNode = queue.poll();
            for (Node n : currNode.neighbors) {
                if (!map.containsKey(n)) {
                    map.put(n, new Node(n.val));
                    queue.offer(n);
                }

                map.get(currNode).neighbors.add(map.get(n));
            }
        }
        return newNode;
    }
}

File: leetcode/_133_.py

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        lookup = {}
        return self.dfs(node, lookup)

    def dfs(self, node, lookup):
        if not node:
            return None
        if node in lookup:
            return lookup[node]

        new_node = Node(node.val)
        lookup[node] = new_node
        for n in node.neighbors:
            new_node.neighbors.append(self.dfs(n, lookup))

        return new_node